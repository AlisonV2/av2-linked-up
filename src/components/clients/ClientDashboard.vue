<template>
  <div class="row">
    <app-title mode="title-img">Your profile</app-title>
    <div class="toast-success" v-if="showSuccessToast">Update successful</div>
    <div class="toast-error" v-if="showErrorToast">
      Oops... Something went wrong :(
    </div>
    <div class="col-12 col-lg-4 mt-4">
      <div class="mt-3">
        <input type="file" @change="setPreviewImage($event)" />
        <div class="error">{{ fileErr }}</div>
      </div>
      <img
        :src="
          previewImage
            ? previewImage
            : require('@/assets/img/default-placeholder.png')
        "
        width="250"
        height="250"
        class="img-fluid"
      />
      <div class="btn-center mt-3 thumbnail-button" @click="setClientThumbnail">
        <app-button mode="save-btn">Save</app-button>
      </div>
      <!-- <div class="error">{{ fileError }}</div> -->
    </div>
    <form class="col-12 col-lg-8 mt-4" @submit.prevent="setClientProfile">
      <div class="form-floating mb-3">
        <input
          type="text"
          class="form-control"
          placeholder="Public Name"
          v-model="profile.name"
        />
        <label>Public Name</label>
      </div>
      <div class="form-floating mb-3">
        <input
          type="textarea"
          class="form-control"
          placeholder="Description "
          v-model="profile.description"
        />
        <label>Description</label>
      </div>
      <div class="input-group">
        <div class="form-floating mb-3 col-lg-6">
          <input
            type="text"
            class="form-control"
            placeholder="City"
            v-model="profile.city"
          />
          <label>City</label>
        </div>
        <div class="form-floating mb-3 col-lg-6">
          <input
            type="text"
            class="form-control"
            placeholder="Zip Code"
            v-model="profile.zip"
          />
          <label>Zip Code</label>
        </div>
      </div>
      <div>
        <select class="form-select" v-model="profile.tattooed">
          <option disabled value="">Already tattooed?</option>
          <option value="yes">Yes</option>
          <option value="no">No</option>
        </select>
      </div>
      <div class="col-12 btn-right mt-3">
        <div class="btn-group mb-3">
          <app-button>Save</app-button>
        </div>
      </div>
    </form>
  </div>
</template>

<script>
import * as Sentry from '@sentry/vue';

/**
 * @exports ClientDashboard
 * @type {Component}
 * @requires Sentry
 * @vue-data {array} previewImage
 * @vue-data {object} profile
 * @vue-data {boolean} showSuccessToast
 * @vue-data {boolean} showErrorToast
 * @vue-data {array} fileSelected
 * @vue-data {string} fileErr
 * @vue-event {object} getClientProfile
 * @vue-event {string} setPreviewImage
 * @vue-event {string} setClientProfile
 * @vue-event {string} setClientThumbnail
 */
export default {
  name: 'ClientDashboard',
  data() {
    return {
      previewImage: null,
      profile: {
        name: '',
        description: '',
        city: '',
        zip: '',
        tattooed: '',
        thumbnail: ''
      },
      showSuccessToast: false,
      showErrorToast: false,
      fileSelected: null,
      fileErr: null,
    };
  },
  async created() {
    await this.$store
      .dispatch('getClientProfile')
      .then(() => {
        this.profile = this.$store.state.clientProfile.clientProfile;
      })
      .then(() => {
        this.previewImage = this.profile.thumbnail;
      })
      .catch((err) => Sentry.captureException(err));
  },
  methods: {
    setPreviewImage($event) {
      const file = $event.target.files[0];
      const types = ['image/png', 'image/jpeg'];

      if (file && types.includes(file.type)) {
        this.fileSelected = file;
        this.fileErr = null;
        this.previewImage = URL.createObjectURL(file);
      } else {
        this.fileSelected = null;
        this.fileErr = 'Please select an image file (png or jpeg).';
      }
    },
    setClientThumbnail() {
      this.$store
        .dispatch('setClientThumbnail', this.fileSelected)
        .then(() => {
          this.showSuccess();
          window.location.reload();
        })
        .catch((err) => {
          this.showError(err);
        });
    },
    async setClientProfile() {
      const profileData = {
        name: this.profile.name,
        description: this.profile.description,
        city: this.profile.city,
        zip: this.profile.zip,
        tattooed: this.profile.tattooed
      };
      await this.$store
        .dispatch('setClientProfile', profileData)
        .then(() => {
          this.showSuccess();
          window.location.reload();
        })
        .catch((err) => {
          this.showError(err);
        });
    },
    showSuccess() {
      this.showSuccessToast = true;
      this.showErrorToast = false;
      setTimeout(() => {
        this.showSuccessToast = false;
      }, 3000);
    },
    showError(err) {
      this.showErrorToast = true;
      this.showSuccessToast = false;
      Sentry.captureException(err);
      setTimeout(() => {
        this.showErrorToast = false;
      }, 3000);
    },
  },
};
</script>
