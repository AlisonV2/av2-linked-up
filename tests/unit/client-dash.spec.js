import { shallowMount } from '@vue/test-utils';
import { createStore } from 'vuex';
import ClientDashboard from '@/components/clients/ClientDashboard';
import AppTitle from '@/components/app/Title';

const actions = {
  getClientProfile: jest.fn(),
  setClientThumbnail: jest.fn(),
  setClientProfile: jest.fn(),
};

const state = {
  clientprofile: {
    clientProfile: {
      name: 'Alison',
      description: 'Unicorn power',
      city: 'Bordeaux',
      zip: '33800',
      tattooed: 'Yes',
      thumbnail: 'url',
    },
  },
};

const store = createStore({
  state,
  actions,
});

const $router = {
  push: jest.fn(),
};

const $route = {
  params: { id: '1' },
};

const config = {
  data() {
    return {
      previewImage: null,
      profile: {
        name: 'Alison',
        description: 'Unicorn power',
        city: 'Bordeaux',
        zip: '33800',
        tattooed: 'Yes',
        thumbnail: 'url',
      },
      showSuccessToast: false,
      showErrorToast: false,
      fileSelected: null,
      fileErr: null,
    };
  },
  global: {
    plugins: [store],
    components: {
      'app-title': AppTitle,
    },
    mocks: {
      $router,
      $route,
    },
  },
};
/**
 * @module ClientDashboardTest
 */
describe('ClientDashboard.vue', () => {
  it('ClientDashboard Snapshot', () => {
    const component = shallowMount(ClientDashboard, config);
    expect(component.element).toMatchSnapshot();
  });

  it('Check if getClientProfile action is called', async () => {
    const component = shallowMount(ClientDashboard, config);
    ClientDashboard.created.call(component.vm);
    await component.vm.$nextTick();
    expect(actions.getClientProfile).toHaveBeenCalled;
  });

  it('Check if setClientProfile action is called', async () => {
    const component = shallowMount(ClientDashboard, config);
    component.find('form').trigger('submit.prevent');
    await component.vm.$nextTick();
    expect(actions.setClientProfile).toHaveBeenCalled;
  });

  it('Check if setClientThumbnail action is called', async () => {
    const component = shallowMount(ClientDashboard, config);
    component.find('.thumbnail-button').trigger('click');
    await component.vm.$nextTick();
    expect(actions.setClientThumbnail).toHaveBeenCalled;
  });
});
