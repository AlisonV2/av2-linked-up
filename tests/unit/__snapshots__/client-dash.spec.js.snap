// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ClientDashboard.vue ClientDashboard Snapshot 1`] = `
<div
  class="row"
>
  <app-title-stub />
  <!--v-if-->
  <!--v-if-->
  <div
    class="col-12 col-lg-4 mt-4"
  >
    <div
      class="mt-3"
    >
      <input
        type="file"
      />
      <div
        class="error"
      />
    </div>
    <img
      alt="Profile Image"
      class="img-fluid"
      height="250"
      src=""
      width="250"
    />
    <div
      class="btn-center mt-3 thumbnail-button"
    >
      <app-button
        mode="save-btn"
      >
        Save
      </app-button>
    </div>
    <div
      class="error"
    />
  </div>
  <form
    class="col-12 col-lg-8 mt-4"
  >
    <div
      class="form-floating mb-3"
    >
      <input
        class="form-control"
        placeholder="Public Name"
        type="text"
      />
      <label>
        Public Name
      </label>
    </div>
    <div
      class="form-floating mb-3"
    >
      <input
        class="form-control"
        placeholder="Description "
        type="textarea"
      />
      <label>
        Description
      </label>
    </div>
    <div
      class="input-group"
    >
      <div
        class="form-floating mb-3 col-lg-6"
      >
        <input
          class="form-control"
          placeholder="City"
          type="text"
        />
        <label>
          City
        </label>
      </div>
      <div
        class="form-floating mb-3 col-lg-6"
      >
        <input
          class="form-control"
          placeholder="Zip Code"
          type="text"
        />
        <label>
          Zip Code
        </label>
      </div>
    </div>
    <div>
      <select
        class="form-select"
      >
        <option
          disabled=""
          value=""
        >
          Already tattooed?
        </option>
        <option
          value="yes"
        >
          Yes
        </option>
        <option
          value="no"
        >
          No
        </option>
      </select>
    </div>
    <div
      class="col-12 btn-right mt-3"
    >
      <div
        class="btn-group mb-3"
      >
        <app-button>
          Save
        </app-button>
      </div>
    </div>
  </form>
</div>
`;
